### ISSUE TITLE
Service URL Truncated

### ISSUE THREAD

**Issue by herringm** [apereo/dotnet-cas-client#18]
We have a third party application that uses URLs with an encoded data parameter. These are used for registration and other purposes and following these links prior to authentication result in any URL encoded characters at the end of the URL being lost during the CAS login process. See the URLs below from the resulting login chain which were captured from the browser developer tools:

```
https://login.denison.edu/cas/login?service=https%3a%2f%2fevents.denison.edu%2fMasterCalendar%2fManageSubscription.aspx%3fdata%3d7J%2fMlsKOh99bxaaUpZSeJRD0avGuP%2brv%2fRo48w4%2brHTk%2bSQ628pOYLZ0zyUzj2M7TwhZdt6mrtyb7419rMmsSQ%3d%3d

https://events.denison.edu/MasterCalendar/ManageSubscription.aspx?data=7J/MlsKOh99bxaaUpZSeJRD0avGuP+rv/Ro48w4+rHTk+SQ628pOYLZ0zyUzj2M7TwhZdt6mrtyb7419rMmsSQ==&ticket=ST-1261537-bwgiAsqkfoewNA7tmuOS-login.denison.edu

https://events.denison.edu/MasterCalendar/ManageSubscription.aspx?data=7J/MlsKOh99bxaaUpZSeJRD0avGuP%20rv/Ro48w4%20rHTk%20SQ628pOYLZ0zyUzj2M7TwhZdt6mrtyb7419rMmsSQ
```

As you'll note when the ticket was removed from the URL so were the two equal signs (=) which resulted in the application failing to load the correct subscription.

If the user previously authenticated; clicking and following this link (which is delivered via an email) works as expected. It only appears to happen when the CAS client parses the URL to retrieve the CAS ticket.

Any help or guidance is appreciated!

**Comment by hokiecoder** [comment]
@scottt732 I have built the assembly using your `url-encoding-bugfix` branch fixes and found that there is still an unresolved issue when no valid ticket exists and the module redirects the client to the CAS server for authentication. Using your fix and a simple query string field value of a plus character ('+'):

1. Make sure there is no currently existing login with the CAS server.
2. Navigate to `https://example.server.com/default.asxp?data=%2B`.
3. Client is redirected to `https://cas.server.com/cas/login?TARGET=https%3a%2f%2fexample.server.com%2fdefault.aspx%3fdata%3d%2b`.
4. Notice how the value of the `data` field is still `%2B` instead of being double URL encoded to `%252B`.

I narrowed down this issue to `UrlUtils.cs`. In `ConstructServiceUrl()`, [line 101](https://github.com/apereo/dotnet-cas-client/blob/a100ed07bcf901a8c351abadb287c28b9f344f18/DotNetCasClient/Utils/UrlUtil.cs#L101), when the 'QueryString' `NameValueCollection` is added to the service URL, it is URL decoded instead of being left as is. I'm not sure exactly where, but it seems to be happening somewhere in the `EnhancedUriBuilder` class. Since the query string parameters are being decoded, they end up being fully decoded once the client redirects after the authentication and validation steps are complete. Thus, in my example above, `data` would end up becoming a space since a plus character is treated as a space.

Instead of the current line 101 in `UrlUtil.cs` I'd recommend the following change:
```
foreach(string name in request.QueryString.AllKeys)
{
    ub.QueryItems.Add(HttpUtility.UrlEncode(name), HttpUtility.UrlEncode(request.QueryString[name]));
}
```
I'm not 100% sure if the field names need to be URL encoded, or more specifically if the query string spec allows for it, but in my testing Google Chrome allowed it.  With this fix, any query string is preserved regardless of the authentication roundtrip happening.

**Comment by serac** [comment]
@hokiecoder can you please fork @scottt732's `url-encoding-bugfix` branch and make the change you suggested above? I'd be happy to review the full patch in context.

**Comment by phantomtypist** [comment]
@serac agreed, I smell unit tests coming! :D

@hokiecoder see the question I just added in PR #53.  If you don't have the time to actually work on the unit test project that's fine, but I'd still like to hear your opinion on the question I asked since we'll be working on this project together in the future.

**Comment by phantomtypist** [comment]
Fix was merged in #53.
